moc_openshift.py

```diff
@@ -469,3 +541,108 @@ def update_rolebindings(self, project_name, role, rolebindings_json):
                 payload["metadata"][key] = rolebindings_json["metadata"][key]
         self.logger.debug("payload -> 2: " + json.dumps(payload))
         return self.put_request(url, payload, True)
+
+    def get_moc_quota(self, project_name):
+        quota_def = self.get_quota_definitions()
+        quota = {}
+        for quota_name in quota_def:
+            quota[quota_name] = quota_def[quota_name]["value"]
+
+        quota_object = {
+            "Version": "0.9",
+            "Kind": "MocQuota",
+            "ProjectName": project_name,
+            "Quota": quota,
+        }
+        return quota_object
+
+    def create_shift_quotas(self, project_name, quota_spec):
+        quota_def = {}
+        # separate the quota_spec by quota_scope
+        for mangled_quota_name in quota_spec:
+            (scope, quota_name) = self.split_quota_name(mangled_quota_name)
+            if scope not in quota_def:
+                quota_def[scope] = {}
+            quota_def[scope][quota_name] = quota_spec[mangled_quota_name]
+        # create the openshift quota resources
+        quota_create_status_code = 200
+        quota_create_msg = ""
+        for scope, quota_item in quota_def.items():
+            resource_quota_json = {
+                "apiVersion": "v1",
+                "kind": "ResourceQuota",
+                "metadata": {"name": f"{project_name.lower()}-{scope.lower()}"},
+                "spec": {"hard": {}},
+            }
+            if scope != "Project":
+                resource_quota_json["spec"]["scopes"] = []
+                resource_quota_json["spec"]["scopes"].append(scope)
+            non_null_quota_count = 0
+            for quota_name in quota_item:
+                if quota_item[quota_name] is not None:
+                    non_null_quota_count += 1
+                    resource_quota_json["spec"]["hard"][quota_name] = quota_item[
+                        quota_name
+                    ]["value"]
+            if non_null_quota_count > 0:
+                url = (
+                    f"{self.get_url()}/api/v1/namespaces/{project_name}/resourcequotas"
+                )
+                resp = self.post_request(url, resource_quota_json, True)
+                time.sleep(2)
+                if resp.status_code in [200, 201]:
+                    quota_create_msg = f"{quota_create_msg} Quota {project_name}/{quota_name} successfully created\n"
+                else:
+                    if resp.status_code > quota_create_status_code:
```

Comment:
The logic here for managing the value of `quota_create_status_code` and `quota_create_msg` seems a little convoluted. In the event of a failure (`resp.status_code` not in `[200, 201]`), couldn't you just return immediately with an error message? I see this pattern elsewhere in the code and I wonder if there's a way to make the behavior a little more straightforward.

Reply:
The basic problem is that we have 1 call to create an MOC Quota that becomes up to 5 calls to OpenShift to create the appropriate resource quotas.  if they all succeed, then we should pass back success, however if one of the 5 fails, then we should pass back failure.  I also want the diagnostic message to pass back information on the 5 responses.

Currently, with the sleep in between calls, I haven't seen any failures.  In the next part, I'll be able to remove the timing as I'll be able to check for the existence of the quota objects as they are created.